client/allocrunner/taskrunner/script_check_hook_test.go

```diff
@@ -0,0 +1,213 @@
+// timeout is reached.
+func TestScript_Exec_TimeoutBasic(t *testing.T) {
+	t.Parallel()
+	exec, cancel := newBlockingScriptExec()
+	defer cancel()
+
+	logger := testlog.HCLogger(t)
+	hb := newFakeHeartbeater()
+	script := newScriptMock(hb, exec, logger, time.Hour, time.Second)
+
+	handle := script.run()
+	defer handle.cancel() // cleanup
+	<-exec.running        // wait until Exec is called
+
+	// Check for UpdateTTL call
+	select {
+	case update := <-hb.heartbeats:
+		require.Equal(t, update.output, context.DeadlineExceeded.Error())
+		require.Equal(t, update.status, api.HealthCritical)
+	case <-time.After(3 * time.Second):
+		t.Fatalf("timed out waiting for script check to exit")
+	}
+
+	// The underlying ScriptExecutor (newBlockScriptExec) *cannot* be
+	// canceled. Only a wrapper around it obeys the context cancelation.
+	require.NotEqual(t, atomic.LoadInt32(&exec.exited), 1,
+		"expected script executor to still be running after timeout")
+
+	// Cancel and watch for exit
+	handle.cancel()
+	select {
+	case <-handle.wait(): // ok!
+	case update := <-hb.heartbeats:
+		t.Errorf("unexpected UpdateTTL call on exit with status=%q", update)
+	case <-time.After(3 * time.Second):
+		t.Fatalf("timed out waiting for script check to exit")
+	}
+}
+
+// TestScript_Exec_TimeoutCritical asserts a script will be killed when
+// the timeout is reached and always set a critical status regardless of what
+// Exec returns.
+func TestScript_Exec_TimeoutCritical(t *testing.T) {
+	t.Parallel()
+	logger := testlog.HCLogger(t)
+	hb := newFakeHeartbeater()
+	script := newScriptMock(hb, sleeperExec{}, logger, time.Hour, time.Nanosecond)
+
+	handle := script.run()
+	defer handle.cancel() // cleanup
+
+	// Check for UpdateTTL call
+	select {
+	case update := <-hb.heartbeats:
+		require.Equal(t, update.output, context.DeadlineExceeded.Error())
+		require.Equal(t, update.status, api.HealthCritical)
+	case <-time.After(3 * time.Second):
+		t.Fatalf("timed out waiting for script check to timeout")
+	}
+}
+
+// TestScript_Exec_Shutdown asserts a script will be executed once more
+// when told to shutdown.
+func TestScript_Exec_Shutdown(t *testing.T) {
+	shutdown := make(chan struct{})
+	exec := newSimpleExec(0, nil)
+	logger := testlog.HCLogger(t)
+	hb := newFakeHeartbeater()
+	script := newScriptMock(hb, exec, logger, time.Hour, 3*time.Second)
+	script.shutdownCh = shutdown
+
+	handle := script.run()
+	defer handle.cancel() // cleanup
+	close(shutdown)       // tell scriptCheck to exit
+
+	select {
+	case update := <-hb.heartbeats:
+		require.Equal(t, update.output, "code=0 err=<nil>")
+		require.Equal(t, update.status, api.HealthPassing)
+	case <-time.After(3 * time.Second):
+		t.Fatalf("timed out waiting for script check to exit")
+	}
+
+	select {
+	case <-handle.wait(): // ok!
+	case <-time.After(3 * time.Second):
+		t.Fatalf("timed out waiting for script check to exit")
+	}
+}
+
+func TestScript_Exec_Codes(t *testing.T) {
```

Comment:
No a blocker but in the future add a comment to tests of the form:
```suggestion
// TestScript_Exec_Codes asserts script exit codes are translated to their
// corresponding Consul health check status.
func TestScript_Exec_Codes(t *testing.T) {
```

Reply:
Done in 2eefcc1